### PR TITLE
Reject responses with explicit set-cookie header

### PR DESCRIPTION
**Pull request by geeksilva97**
This PR addresses the issue #1619 .

## Proposed changes

- reject response with explicit `<<"set-cookie">>` in headers

### CODE REVIEW

File: src/cowboy_req.erl [hunk 1]
```diff
@@ -781,6 +788,9 @@ inform(Status, Req) ->
 inform(_, _, #{has_sent_resp := _}) ->
 	exit({response_error, response_already_sent,
 		'The final response has already been sent.'});
+inform(_, _, #{<<"set-cookie">> := _}) ->
```
Comment (geeksilva97) [inline src/cowboy_req.erl:791]:
```suggestion
inform(_, #{<<"set-cookie">> := _}, _) ->
```

**Comment by geeksilva97** [comment]
Could you take a look when you have time, @essen ? Thanks in advance.

File: src/cowboy_req.erl [hunk 2]
```diff
@@ -15,6 +15,9 @@
 
 -module(cowboy_req).
 
+-define(INVALID_COOKIE_HEADER_ERROR, exit({response_error, invalid_header,
+    'The set-cookie header is special and must be set using cowboy_req:set_resp_cookie/3,4.'})).
```
Comment (essen) [inline src/cowboy_req.erl:19]: No macros please.

**Comment by essen** [comment]
Yes tests are fine like this, though since there's a crash you should also call `ct_helper:ignore(M, F, A),` before the crash, with MFA where the crash occurs. It will silence the expected crash.

File: src/cowboy_req.erl (hunk 2 above)
Comment (geeksilva97) [reply src/cowboy_req.erl:19]: for sure. will move to a function

Comment (essen) [reply src/cowboy_req.erl:19]: Just inline is fine. There isn't too many of them. The messages are likely to be slightly edited later on anyway depending on context.

**Comment by geeksilva97** [comment]
@essen would you have any clue how we can test this on `stream_trailers`?

I tried adding this to `resp_h.erl` - inside of `do(<<"stream_trailers">>, Req0, Opts)`.

```erl
<<"set_cookie">> ->
			ct_helper:ignore(cowboy_req, stream_trailers, 2),
			Req = cowboy_req:stream_reply(200, #{
				<<"trailer">> => <<"grpc-status">>
			}, Req0),
			%% The size should be larger than StreamSize and ConnSize
			cowboy_req:stream_body(<<0:80000000>>, nofin, Req),
			cowboy_req:stream_trailers(#{
				<<"set-cookie">> => <<"name=cormano">>
			}, Req),
			{ok, Req, Opts};
```

Then I tried something like this

```erl
do_trailers("resp/stream_trailers/set_cookie", Config)
```

It broke the test suite - I didn't even check the return

It breaks somewhere else it is also a badmatch.

**Comment by essen** [comment]
Sometimes it can be difficult to test crashes. Some of the Cowboy tests trigger the crash but the test case has no way of confirming it happened. That's OK for the time being.

I will pick this up after I'm done with my current ticket. A quick look seems fine. Thank you!

**Comment by geeksilva97** [comment]
> Sometimes it can be difficult to test crashes. Some of the Cowboy tests trigger the crash but the test case has no way of confirming it happened. That's OK for the time being.
> 
> I will pick this up after I'm done with my current ticket. A quick look seems fine. Thank you!

Sounds good. Thanks for the help. Let me know if I can help with something. 

I will look for another issue - would you have any suggestion?

**Comment by essen** [comment]
Merged with a number of fixes, thank you!

Note that most of the tests were not testing the right crash as you copied test code that used bad arguments to trigger crashes and didn't update them.

**Comment by josevalim** [comment]
Hi @essen! This pull request unfortunately breaks all Elixir projects that depend on Cowboy via Plug, since Plug manages the cookie headers by itself. Is there any chance this could be skipped by using a flag? Or alternatively, would you accept a pull request that adds `set_resp_cookie(HeaderValue, Req :: cowboy_req:req())`, so we can give the whole header value (which we have already accumulated and returned)? Thank you.

**Comment by essen** [comment]
For now you can modify the Req object directly (basically doing what `set_resp_cookie(HeaderValue, ...)` would do).

Not sure this has to be in Cowboy but feel free to open a PR and let's gather feedback.

**Comment by essen** [comment]
For future releases perhaps Plug can add Cowboy master to its CI? Maybe on a schedule, running every week. This would have catched this. I am working immediately on 2.12 and this will take a while (HTTP/3) so getting feedback before release would help.

**Pull request closed without merging**